server/src/main/java/org/elasticsearch/cluster/coordination/PreVoteCollectorFactory.java

```diff
@@ -0,0 +1,204 @@
+        }
+    }
+
+    // package-private for testing, actual usages TODO
+    long getMaxTermSeen() {
+        return maxTermSeen.get();
+    }
+
+    @Override
+    public String toString() {
+        return "PreVoteCollectorFactory{" +
+            "currentLeader=" + currentLeader +
+            ", preVoteResponse=" + preVoteResponse +
+            '}';
+    }
+
+    private class PreVotingRound implements Releasable {
+        private final Set<DiscoveryNode> preVotesReceived = newConcurrentSet();
+        private final AtomicBoolean electionStarted = new AtomicBoolean();
+        private final PreVoteRequest preVoteRequest;
+        private final ClusterState clusterState;
+        private final AtomicBoolean isRunning = new AtomicBoolean();
+
+        PreVotingRound(ClusterState clusterState) {
+            this.clusterState = clusterState;
+            preVoteRequest = new PreVoteRequest(transportService.getLocalNode(), preVoteResponse.getCurrentTerm());
+        }
+
+        void start(final Iterable<DiscoveryNode> broadcastNodes) {
+
+            final boolean isRunningChanged = isRunning.compareAndSet(false, true);
+            assert isRunningChanged;
+
+            broadcastNodes.forEach(n -> transportService.sendRequest(n, REQUEST_PRE_VOTE_ACTION_NAME, preVoteRequest,
+                new TransportResponseHandler<PreVoteResponse>() {
+                    @Override
+                    public void handleResponse(PreVoteResponse response) {
+                        handlePreVoteResponse(response, n);
+                    }
+
+                    @Override
+                    public void handleException(TransportException exp) {
+                        if (exp.getRootCause() instanceof CoordinationStateRejectedException) {
+                            logger.debug("{} failed: {}", this, exp.getRootCause().getMessage());
+                        } else {
+                            logger.debug(new ParameterizedMessage("{} failed", this), exp);
+                        }
+                    }
+
+                    @Override
+                    public String executor() {
+                        return Names.GENERIC;
+                    }
+
+                    @Override
+                    public String toString() {
+                        return "TransportResponseHandler{" + PreVoteCollectorFactory.this + ", node=" + n + '}';
+                    }
+                }));
+        }
+
+        private void handlePreVoteResponse(PreVoteResponse response, DiscoveryNode sender) {
+            if (isRunning.get() == false) {
+                logger.debug("{} ignoring {} from {}, no longer running", this, response, sender);
+                return;
+            }
+
+            final long currentMaxTermSeen = updateMaxTermSeen(response.getCurrentTerm());
+
+            final PreVoteResponse currentPreVoteResponse = preVoteResponse;
+            if (response.getLastAcceptedTerm() > currentPreVoteResponse.getLastAcceptedTerm()
```

Comment:
I wonder if we can do something similar as for the `isElectionQuorum` trick where we share implementation of these checks with `CoordinationState`.

Reply:
I'd prefer not to do so. The machinery to share these few lines of code obscures what they do, and the extra abstraction in `CoordinationState` takes it further away from the formal model.